src/basic/filesystem-sets.py

```diff
@@ -0,0 +1,352 @@
+        "xfs",
+    ),
+    (
+        "@historical-block",
+        "Historical block device filesystems",
+        "ext2",
+        "ext3",
+        "minix",
+    ),
+    (
+        "@network",
+        "Well-known network filesystems",
+        "afs",
+        "ceph",
+        "cifs",
+        "gfs",
+        "gfs2",
+        "ncp",
+        "ncpfs",
+        "nfs",
+        "nfs4",
+        "ocfs2",
+        "orangefs",
+        "pvfs2",
+        "smb3",
+        "smbfs",
+    ),
+    (
+        "@privileged-api",
+        "Privileged filesystem API",
+        "bpf",
+        "debugfs",
+        "pstore",
+        "tracefs",
+    ),
+    (
+        "@security",
+        "Security/MAC API VFS",
+        "apparmorfs",
+        "selinuxfs",
+        "smackfs",
+    ),
+    (
+        "@temporary",
+        "Temporary filesystems",
+        "ramfs",
+        "tmpfs",
+    ),
+    (
+        "@known",
+        "All known filesystems declared in the kernel",
+        *NAME_TO_MAGIC.keys(),
+    ),
+]
+
+def generate_gperf():
+    print("""\
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+%{
+#if __GNUC__ >= 15
+_Pragma("GCC diagnostic ignored \\"-Wzero-as-null-pointer-constant\\"")
+#endif
+#include <linux/magic.h>
+
+#include "filesystems.h"
+#include "stat-util.h"
+
+struct FilesystemMagic {
+        const char *name;
+        statfs_f_type_t magic[FILESYSTEM_MAGIC_MAX];
+};
+%}
+struct FilesystemMagic;
+%language=ANSI-C
+%define hash-function-name filesystems_gperf_hash
+%define lookup-function-name filesystems_gperf_lookup
+%define slot-name name
+%readonly-tables
+%omit-struct-type
+%struct-type
+%includes
+%%""")
+    for name, magics in NAME_TO_MAGIC.items():
+        print(f"{name + ',':16} {{{', '.join(magics)}}}")
+
+def generate_fs_type_to_string():
+    print("""\
+#include <linux/magic.h>
+#include "filesystems.h"
+
+const char* fs_type_to_string(statfs_f_type_t magic) {
+        switch (magic) {""")
+
+    for name, magics in NAME_TO_MAGIC.items():
+        if name in OBSOLETE_NAMES:
+            continue
+        for magic in magics:
+            print(f'        case (statfs_f_type_t) {magic}:')
+        print(f'                return "{name}";')
+
+    print("""\
+        }
+        return NULL;
+}""")
+
+def generate_filesystem_sets():
+    print('const FilesystemSet filesystem_sets[_FILESYSTEM_SET_MAX] = {')
+
+    for name, desc, *filesystems in FILESYSTEM_SETS:
+        enum = 'FILESYSTEM_SET_' + name[1:].upper().replace('-', '_')
+
+        print(f'        [{enum}] = {{')
+        print(f'                .name = "{name}",')
+        print(f'                .help = "{desc}",')
+        print(f'                .value =')
+        for filesystem in filesystems:
+            print(f'                "{filesystem}\\0"')
+        print('        },')
+
+    print('};')
+
+def magic_defines():
+    cpp = os.environ['CPP'].split()
+    out = subprocess.check_output(
+        [*cpp, '-dM', '-include', 'linux/magic.h', '-'],
```

Comment:
This requires the latest kernel header is installed on the build system.
Please pass `system_include_args` to the script.

Reply:
OK, that makes sense to me.